salt/states/zfs.py

```diff
@@ -706,4 +706,217 @@ def promoted(name):
 
     return ret
 
+
+def scheduled_snapshot(name, prefix, recursive=True, schedule=None):
+    '''
+    create snapshots based on a schedule
+
+    name : string
+        name of filesystem or volume
+    prefix : string
+        prefix for the snapshots
+        e.g. 'test' will result in snapshots being named 'test-YYYYMMDD_HHMM'
+    recursive : boolean
+        create snapshots for all children also
+    schedule : dict
+        dict holding the schedule, the following keys are available (minute, hour,
+        day, month, and year) by default all are set to 0 the value indicated the
+        number of snapshots of that type to keep around.
+
+    ..warning::
```

Comment:
Hello minion, please keep 7 snapshots of type day and also keep 12 snapshots of type month.

After every run, the minion will have (maximum) 7 snapshots of type day and 12 of type month.
If the system state changes, it will make sure there are only 7 and 12 snapshots of the given type respectively by either pruning some of the older ones or if there are not enough by slowly adding them until the state requested is kept.

That seems idempotent to me. Perhaps the description could be somewhat improved.
The state simply declares what the desired number of snapshots of a certain type is for a dataset and will ensure we are at the desired state or will make it the desired state. (Although if there are less than the desired state it could take a few runs to get there)